File: typescript/tests/unit/agent-bay-pause-resume.test.ts
import { AgentBay, Session } from "../../src";
import * as sinon from "sinon";
import { SessionPauseResult, SessionResumeResult } from "../../src/types/api-response";

// Mock data for tests
const mockSessionData = {
  sessionId: "mock-session-123",
  resourceUrl: "mock-resource-url",
};

describe("AgentBay Pause and Resume", () => {
  let mockClient: any;
  let pauseSessionAsyncStub: sinon.SinonStub;
  let resumeSessionAsyncStub: sinon.SinonStub;
  let getSessionStub: sinon.SinonStub;
  let agentBay: AgentBay;
  let session: Session;

  beforeEach(() => {
    // Create mock client
    mockClient = {
      pauseSessionAsync: sinon.stub(),
      resumeSessionAsync: sinon.stub(),
      getSession: sinon.stub(),
    };

    // Create AgentBay instance with mock client
    agentBay = new AgentBay({ apiKey: "test-api-key" });
    (agentBay as any).client = mockClient;

    // Create mock session
    session = new Session(agentBay, mockSessionData.sessionId);

    // Set up stubs
    pauseSessionAsyncStub = mockClient.pauseSessionAsync;
    resumeSessionAsyncStub = mockClient.resumeSessionAsync;
    getSessionStub = mockClient.getSession;
  });

  afterEach(() => {
    sinon.restore();
  });

  describe("pauseAsync", () => {
    it("should successfully pause a session", async () => {
      // Mock the PauseSessionAsync response
      const mockPauseResponse = {
        body: {
          success: true,
          code: "Success",
          message: "Session pause initiated successfully",
          requestId: "pause-request-id",
          httpStatusCode: 200,
        },
        to_map: () => ({
          body: {
            success: true,
            code: "Success",
            message: "Session pause initiated successfully",
            requestId: "pause-request-id",
            httpStatusCode: 200,
          }
        })
      };

      // Mock get_session to return PAUSED immediately
      const mockGetSessionResponse = {
        body: {
          success: true,
          data: {
            sessionId: mockSessionData.sessionId,
            status: "PAUSED",
          }
        },
        requestId: "get-session-request-id"
      };

      pauseSessionAsyncStub.resolves(mockPauseResponse);
      getSessionStub.resolves(mockGetSessionResponse);

      // Call the method
      const result: SessionPauseResult = await agentBay.pauseAsync(session);

      // Verify the result
      expect(result).toBeDefined();
      expect(result.success).toBe(true);
      expect(result.requestId).toBe("get-session-request-id");
      expect(result.status).toBe("PAUSED");
      expect(result.errorMessage).toBeUndefined();

      // Verify that the API was called with correct parameters
      expect(pauseSessionAsyncStub.calledOnce).toBe(true);
      const calledWith = pauseSessionAsyncStub.getCall(0).args[0];
      expect(calledWith.sessionId).toBe(mockSessionData.sessionId);
    });

    it("should handle pause API error", async () => {
      // Mock the PauseSessionAsync response with error
      const mockPauseResponse = {
        body: {
          success: false,
          code: "InvalidSession",
          message: "Session not found",
          requestId: "test-request-id",
          httpStatusCode: 404,
        },
        to_map: () => ({
          body: {
            success: false,
            code: "InvalidSession",
            message: "Session not found",
            requestId: "test-request-id",
            httpStatusCode: 404,
          }
        })
      };

      pauseSessionAsyncStub.resolves(mockPauseResponse);

      // Call the method
      const result: SessionPauseResult = await agentBay.pauseAsync(session);

      // Verify the result
      expect(result).toBeDefined();
      expect(result.success).toBe(false);
      expect(result.requestId).toBe("test-request-id");
      expect(result.errorMessage).toContain("InvalidSession");
      expect(result.code).toBe("InvalidSession");
      expect(result.message).toBe("Session not found");
      expect(result.httpStatusCode).toBe(404);
    });

    it("should handle network error", async () => {
      // Mock network error
      pauseSessionAsyncStub.rejects(new Error("Network error"));

      // Call the method
      const result: SessionPauseResult = await agentBay.pauseAsync(session);

      // Verify the result
      expect(result).toBeDefined();
      expect(result.success).toBe(false);
      expect(result.errorMessage).toContain("Network error");
    });

    it("should handle timeout scenario", async () => {
      // Mock the PauseSessionAsync response
      const mockPauseResponse = {
        body: {
          success: true,
          code: "Success",
          message: "Session pause initiated successfully",
          requestId: "test-request-id",
          httpStatusCode: 200,
        },
        to_map: () => ({
          body: {
            success: true,
            code: "Success",
            message: "Session pause initiated successfully",
            requestId: "test-request-id",
            httpStatusCode: 200,
          }
        })
      };

      // Mock get_session to always return RUNNING (never PAUSED)
      const mockGetSessionResponse = {
        body: {
          success: true,
          data: {
            sessionId: mockSessionData.sessionId,
            status: "RUNNING",
          }
        }
      };

      pauseSessionAsyncStub.resolves(mockPauseResponse);
      getSessionStub.resolves(mockGetSessionResponse);

      // Call the method with short timeout
      const result: SessionPauseResult = await agentBay.pauseAsync(session, 0.1, 0.05);

      // Verify the result
      expect(result).toBeDefined();
      expect(result.success).toBe(false);
      expect(result.errorMessage).toContain("timed out");
    });

    it("should handle get session failure", async () => {
      // Mock the PauseSessionAsync response
      const mockPauseResponse = {
        body: {
          success: true,
          code: "Success",
          message: "Session pause initiated successfully",
          requestId: "test-request-id",
          httpStatusCode: 200,
        },
        to_map: () => ({
          body: {
            success: true,
            code: "Success",
            message: "Session pause initiated successfully",
            requestId: "test-request-id",
            httpStatusCode: 200,
          }
        })
      };

      // Mock get_session failure
      const mockGetSessionResponse = {
        body: {
          success: false,
          code: "GetSessionError",
          message: "Failed to get session status",
          requestId: "test-request-id",
        }
      };

      pauseSessionAsyncStub.resolves(mockPauseResponse);
      getSessionStub.resolves(mockGetSessionResponse);

      // Call the method
      const result: SessionPauseResult = await agentBay.pauseAsync(session);

      // Verify the result
      expect(result).toBeDefined();
      expect(result.success).toBe(false);
      expect(result.errorMessage).toContain("Failed to get session status");
    });
  });

  describe("resumeAsync", () => {
    it("should successfully resume a session", async () => {
      // Mock the ResumeSessionAsync response
      const mockResumeResponse = {
        body: {
          success: true,
          code: "Success",
          message: "Session resume initiated successfully",
          requestId: "resume-request-id",
          httpStatusCode: 200,
        },
        to_map: () => ({
          body: {
            success: true,
            code: "Success",
            message: "Session resume initiated successfully",
            requestId: "resume-request-id",
            httpStatusCode: 200,
          }
        })
      };

      // Mock get_session to return RUNNING immediately
      const mockGetSessionResponse = {
        body: {
          success: true,
          data: {
            sessionId: mockSessionData.sessionId,
            status: "RUNNING",
          }
        },
        requestId: "get-session-request-id"
      };

      resumeSessionAsyncStub.resolves(mockResumeResponse);
      getSessionStub.resolves(mockGetSessionResponse);

      // Call the method
      const result: SessionResumeResult = await agentBay.resumeAsync(session);

      // Verify the result
      expect(result).toBeDefined();
      expect(result.success).toBe(true);
      expect(result.requestId).toBe("get-session-request-id");
      expect(result.status).toBe("RUNNING");
      expect(result.errorMessage).toBeUndefined();

      // Verify that the API was called with correct parameters
      expect(resumeSessionAsyncStub.calledOnce).toBe(true);
      const calledWith = resumeSessionAsyncStub.getCall(0).args[0];
      expect(calledWith.sessionId).toBe(mockSessionData.sessionId);
    });

    it("should handle resume API error", async () => {
      // Mock the ResumeSessionAsync response with error
      const mockResumeResponse = {
        body: {
          success: false,
          code: "InvalidSession",
          message: "Session not found",
          requestId: "test-request-id",
          httpStatusCode: 404,
        },
        to_map: () => ({
          body: {
            success: false,
            code: "InvalidSession",
            message: "Session not found",
            requestId: "test-request-id",
            httpStatusCode: 404,
          }
        })
      };

      resumeSessionAsyncStub.resolves(mockResumeResponse);

      // Call the method
      const result: SessionResumeResult = await agentBay.resumeAsync(session);

      // Verify the result
      expect(result).toBeDefined();
      expect(result.success).toBe(false);
      expect(result.requestId).toBe("test-request-id");
      expect(result.errorMessage).toContain("InvalidSession");
      expect(result.code).toBe("InvalidSession");
      expect(result.message).toBe("Session not found");
      expect(result.httpStatusCode).toBe(404);
    });

    it("should handle network error", async () => {
      // Mock network error
      resumeSessionAsyncStub.rejects(new Error("Network error"));

      // Call the method
      const result: SessionResumeResult = await agentBay.resumeAsync(session);

      // Verify the result
      expect(result).toBeDefined();
      expect(result.success).toBe(false);
      expect(result.errorMessage).toContain("Network error");
    });

    it("should handle timeout scenario", async () => {
      // Mock the ResumeSessionAsync response
      const mockResumeResponse = {
        body: {
          success: true,
          code: "Success",
          message: "Session resume initiated successfully",
          requestId: "test-request-id",
          httpStatusCode: 200,
        },
        to_map: () => ({
          body: {
            success: true,
            code: "Success",
            message: "Session resume initiated successfully",
            requestId: "test-request-id",
            httpStatusCode: 200,
          }
        })
      };

      // Mock get_session to always return PAUSED (never RUNNING)
      const mockGetSessionResponse = {
        body: {
          success: true,
          data: {
            sessionId: mockSessionData.sessionId,
            status: "PAUSED",
          }
        }
      };

      resumeSessionAsyncStub.resolves(mockResumeResponse);
      getSessionStub.resolves(mockGetSessionResponse);

      // Call the method with short timeout
      const result: SessionResumeResult = await agentBay.resumeAsync(session, 0.1, 0.05);

      // Verify the result
      expect(result).toBeDefined();
      expect(result.success).toBe(false);
      expect(result.errorMessage).toContain("timed out");
    });

    it("should handle get session failure", async () => {
      // Mock the ResumeSessionAsync response
      const mockResumeResponse = {
        body: {
          success: true,
          code: "Success",
          message: "Session resume initiated successfully",
          requestId: "test-request-id",
          httpStatusCode: 200,
        },
        to_map: () => ({
          body: {
            success: true,
            code: "Success",
            message: "Session resume initiated successfully",
            requestId: "test-request-id",
            httpStatusCode: 200,
          }
        })
      };

      // Mock get_session failure
      const mockGetSessionResponse = {
        body: {
          success: false,
          code: "GetSessionError",
          message: "Failed to get session status",
          requestId: "test-request-id",
        }
      };

      resumeSessionAsyncStub.resolves(mockResumeResponse);
      getSessionStub.resolves(mockGetSessionResponse);

      // Call the method
      const result: SessionResumeResult = await agentBay.resumeAsync(session);

      // Verify the result
      expect(result).toBeDefined();
      expect(result.success).toBe(false);
      expect(result.errorMessage).toContain("Failed to get session status");
    });
  });
});

File: golang/tests/pkg/unit/agentbay_pause_resume_test.go
package agentbay_test

import (
	"errors"
	"testing"

	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay"
	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay/models"
	"github.com/aliyun/wuying-agentbay-sdk/golang/tests/pkg/unit/mock"
	"github.com/golang/mock/gomock"
	"github.com/stretchr/testify/assert"
)

// TestAgentBay_BetaPause tests AgentBay BetaPause method
func TestAgentBay_BetaPause(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	// Create mock AgentBay client
	mockAgentBay := mock.NewMockAgentBayInterface(ctrl)

	// Create a real session for testing
	client, _ := agentbay.NewAgentBay("test-api-key")
	testSession := agentbay.NewSession(client, "test-session-id")

	// Set expected behavior for successful pause
	expectedResult := &models.SessionPauseResult{
		ApiResponse: models.ApiResponse{
			RequestID: "test-request-id-agentbay",
		},
		Success: true,
		Status:  "PAUSED",
	}
	mockAgentBay.EXPECT().BetaPause(testSession, 600, 2.0).Return(expectedResult, nil)

	// Test Pause method call
	result, err := mockAgentBay.BetaPause(testSession, 600, 2.0)

	// Verify call success
	assert.NoError(t, err)
	assert.NotNil(t, result)
	assert.True(t, result.Success)
	assert.Equal(t, "PAUSED", result.Status)
	assert.Equal(t, "test-request-id-agentbay", result.RequestID)
}

// TestAgentBay_BetaPause_Error tests error handling in AgentBay BetaPause method
func TestAgentBay_BetaPause_Error(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	// Create mock AgentBay client
	mockAgentBay := mock.NewMockAgentBayInterface(ctrl)

	// Create a real session for testing
	client, _ := agentbay.NewAgentBay("test-api-key")
	testSession := agentbay.NewSession(client, "test-session-id")

	// Set expected behavior for error case
	expectedErr := errors.New("agentbay pause operation failed")
	mockAgentBay.EXPECT().BetaPause(testSession, 600, 2.0).Return(nil, expectedErr)

	// Test Pause method call with error
	result, err := mockAgentBay.BetaPause(testSession, 600, 2.0)

	// Verify error handling
	assert.Error(t, err)
	assert.Nil(t, result)
	assert.Equal(t, expectedErr, err)
}

// TestAgentBay_BetaPause_WithCustomParams tests AgentBay BetaPause method with custom parameters
func TestAgentBay_BetaPause_WithCustomParams(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	// Create mock AgentBay client
	mockAgentBay := mock.NewMockAgentBayInterface(ctrl)

	// Create a real session for testing
	client, _ := agentbay.NewAgentBay("test-api-key")
	testSession := agentbay.NewSession(client, "test-session-id")

	// Set expected behavior for successful pause with custom parameters
	expectedResult := &models.SessionPauseResult{
		ApiResponse: models.ApiResponse{
			RequestID: "test-request-id-agentbay-custom",
		},
		Success: true,
		Status:  "PAUSED",
	}
	mockAgentBay.EXPECT().BetaPause(testSession, 300, 1.5).Return(expectedResult, nil)

	// Test Pause method call with custom parameters
	result, err := mockAgentBay.BetaPause(testSession, 300, 1.5)

	// Verify call success
	assert.NoError(t, err)
	assert.NotNil(t, result)
	assert.True(t, result.Success)
	assert.Equal(t, "PAUSED", result.Status)
	assert.Equal(t, "test-request-id-agentbay-custom", result.RequestID)
}

// TestAgentBay_BetaResume tests AgentBay BetaResume method
func TestAgentBay_BetaResume(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	// Create mock AgentBay client
	mockAgentBay := mock.NewMockAgentBayInterface(ctrl)

	// Create a real session for testing
	client, _ := agentbay.NewAgentBay("test-api-key")
	testSession := agentbay.NewSession(client, "test-session-id")

	// Set expected behavior for successful resume
	expectedResult := &models.SessionResumeResult{
		ApiResponse: models.ApiResponse{
			RequestID: "test-request-id-resume-agentbay",
		},
		Success: true,
		Status:  "RUNNING",
	}
	mockAgentBay.EXPECT().BetaResume(testSession, 600, 2.0).Return(expectedResult, nil)

	// Test Resume method call
	result, err := mockAgentBay.BetaResume(testSession, 600, 2.0)

	// Verify call success
	assert.NoError(t, err)
	assert.NotNil(t, result)
	assert.True(t, result.Success)
	assert.Equal(t, "RUNNING", result.Status)
	assert.Equal(t, "test-request-id-resume-agentbay", result.RequestID)
}

// TestAgentBay_BetaResume_Error tests error handling in AgentBay BetaResume method
func TestAgentBay_BetaResume_Error(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	// Create mock AgentBay client
	mockAgentBay := mock.NewMockAgentBayInterface(ctrl)

	// Create a real session for testing
	client, _ := agentbay.NewAgentBay("test-api-key")
	testSession := agentbay.NewSession(client, "test-session-id")

	// Set expected behavior for error case
	expectedErr := errors.New("agentbay resume operation failed")
	mockAgentBay.EXPECT().BetaResume(testSession, 600, 2.0).Return(nil, expectedErr)

	// Test Resume method call with error
	result, err := mockAgentBay.BetaResume(testSession, 600, 2.0)

	// Verify error handling
	assert.Error(t, err)
	assert.Nil(t, result)
	assert.Equal(t, expectedErr, err)
}

// TestAgentBay_BetaResume_WithCustomParams tests AgentBay BetaResume method with custom parameters
func TestAgentBay_BetaResume_WithCustomParams(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	// Create mock AgentBay client
	mockAgentBay := mock.NewMockAgentBayInterface(ctrl)

	// Create a real session for testing
	client, _ := agentbay.NewAgentBay("test-api-key")
	testSession := agentbay.NewSession(client, "test-session-id")

	// Set expected behavior for successful resume with custom parameters
	expectedResult := &models.SessionResumeResult{
		ApiResponse: models.ApiResponse{
			RequestID: "test-request-id-resume-agentbay-custom",
		},
		Success: true,
		Status:  "RUNNING",
	}
	mockAgentBay.EXPECT().BetaResume(testSession, 300, 1.5).Return(expectedResult, nil)

	// Test Resume method call with custom parameters
	result, err := mockAgentBay.BetaResume(testSession, 300, 1.5)

	// Verify call success
	assert.NoError(t, err)
	assert.NotNil(t, result)
	assert.True(t, result.Success)
	assert.Equal(t, "RUNNING", result.Status)
	assert.Equal(t, "test-request-id-resume-agentbay-custom", result.RequestID)
}

// TestAgentBay_BetaPause_UnexpectedStateRunning tests BetaPause method with unexpected RUNNING state
func TestAgentBay_BetaPause_UnexpectedStateRunning(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	// Create mock AgentBay client
	mockAgentBay := mock.NewMockAgentBayInterface(ctrl)

	// Create a real session for testing
	client, _ := agentbay.NewAgentBay("test-api-key")
	testSession := agentbay.NewSession(client, "test-session-id")

	// Set expected behavior for pause with unexpected state
	expectedResult := &models.SessionPauseResult{
		ApiResponse: models.ApiResponse{
			RequestID: "test-request-id",
		},
		Success:      false,
		ErrorMessage: "Session pause failed: unexpected state 'RUNNING'",
		Status:       "RUNNING",
	}
	mockAgentBay.EXPECT().BetaPause(testSession, 600, 2.0).Return(expectedResult, nil)

	// Test Pause method call
	result, err := mockAgentBay.BetaPause(testSession, 600, 2.0)

	// Verify the result
	assert.NoError(t, err)
	assert.NotNil(t, result)
	assert.False(t, result.Success)
	assert.Equal(t, "RUNNING", result.Status)
	assert.Contains(t, result.ErrorMessage, "unexpected state")
	assert.Contains(t, result.ErrorMessage, "RUNNING")
}

// TestAgentBay_Pause_UnexpectedStateOther tests Pause method with unexpected OTHER state
func TestAgentBay_Pause_UnexpectedStateOther(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	// Create mock AgentBay client
	mockAgentBay := mock.NewMockAgentBayInterface(ctrl)

	// Create a real session for testing
	client, _ := agentbay.NewAgentBay("test-api-key")
	testSession := agentbay.NewSession(client, "test-session-id")

	// Set expected behavior for pause with unexpected state
	expectedResult := &models.SessionPauseResult{
		ApiResponse: models.ApiResponse{
			RequestID: "test-request-id",
		},
		Success:      false,
		ErrorMessage: "Session pause failed: unexpected state 'OTHER'",
		Status:       "OTHER",
	}
	mockAgentBay.EXPECT().BetaPause(testSession, 600, 2.0).Return(expectedResult, nil)

	// Test Pause method call
	result, err := mockAgentBay.BetaPause(testSession, 600, 2.0)

	// Verify the result
	assert.NoError(t, err)
	assert.NotNil(t, result)
	assert.False(t, result.Success)
	assert.Equal(t, "OTHER", result.Status)
	assert.Contains(t, result.ErrorMessage, "unexpected state")
	assert.Contains(t, result.ErrorMessage, "OTHER")
}

// TestAgentBay_Resume_UnexpectedStatePaused tests Resume method with unexpected PAUSED state
func TestAgentBay_Resume_UnexpectedStatePaused(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	// Create mock AgentBay client
	mockAgentBay := mock.NewMockAgentBayInterface(ctrl)

	// Create a real session for testing
	client, _ := agentbay.NewAgentBay("test-api-key")
	testSession := agentbay.NewSession(client, "test-session-id")

	// Set expected behavior for resume with unexpected state
	expectedResult := &models.SessionResumeResult{
		ApiResponse: models.ApiResponse{
			RequestID: "test-request-id",
		},
		Success:      false,
		ErrorMessage: "Session resume failed: unexpected state 'PAUSED'",
		Status:       "PAUSED",
	}
	mockAgentBay.EXPECT().BetaResume(testSession, 600, 2.0).Return(expectedResult, nil)

	// Test Resume method call
	result, err := mockAgentBay.BetaResume(testSession, 600, 2.0)

	// Verify the result
	assert.NoError(t, err)
	assert.NotNil(t, result)
	assert.False(t, result.Success)
	assert.Equal(t, "PAUSED", result.Status)
	assert.Contains(t, result.ErrorMessage, "unexpected state")
	assert.Contains(t, result.ErrorMessage, "PAUSED")
}

// TestAgentBay_Resume_UnexpectedStateUnknown tests Resume method with unexpected UNKNOWN state
func TestAgentBay_Resume_UnexpectedStateUnknown(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	// Create mock AgentBay client
	mockAgentBay := mock.NewMockAgentBayInterface(ctrl)

	// Create a real session for testing
	client, _ := agentbay.NewAgentBay("test-api-key")
	testSession := agentbay.NewSession(client, "test-session-id")

	// Set expected behavior for resume with unexpected state
	expectedResult := &models.SessionResumeResult{
		ApiResponse: models.ApiResponse{
			RequestID: "test-request-id",
		},
		Success:      false,
		ErrorMessage: "Session resume failed: unexpected state 'UNKNOWN'",
		Status:       "UNKNOWN",
	}
	mockAgentBay.EXPECT().BetaResume(testSession, 600, 2.0).Return(expectedResult, nil)

	// Test Resume method call
	result, err := mockAgentBay.BetaResume(testSession, 600, 2.0)

	// Verify the result
	assert.NoError(t, err)
	assert.NotNil(t, result)
	assert.False(t, result.Success)
	assert.Equal(t, "UNKNOWN", result.Status)
	assert.Contains(t, result.ErrorMessage, "unexpected state")
	assert.Contains(t, result.ErrorMessage, "UNKNOWN")
}
